proposals/p2200.md

```diff
@@ -0,0 +1,1075 @@
+depends on the value of the parameter, not just its type. As an example, whether
+two array types are compatible depends on whether they have the same size. With
+a symbolic constant sizes, they will only be considered equal if the compiler
+can show that the two sizes are always equal symbolically. If the size is a
+template parameter, the checking will be delayed until the value of the template
+parameter is known.
+
+### Template dependent
+
+Expressions fall under three categories:
+
+-   Expressions that are valid and have meaning determined without knowing the
+    value of any template parameter are _not template dependent_.
+-   Expressions whose meaning and validity requires knowing the value of a
+    template parameter are _template dependent_.
+-   Expressions that have a meaning without knowing the value of any template
+    parameter, assuming it is valid, but whose validity requires knowing the
+    value of a template parameter are _template validity dependent_. These
+    expressions can trigger a monomorphization error, but are not considered
+    template dependent for purposes of a containing expression.
+
+#### Unqualified member name lookup
+
+There are three cases when performing unqualified member-name lookup into a
+templated type:
+
+```carbon
+fn F[template T:! I](x: T) {
+  x.G();
+}
+```
+
+-   If generic name lookup would succeed, in this example it would be because
+    `G` is a member of `I`, then the result of name lookup is template validity
+    dependent. This means that template instantiation may fail if `T` has a
+    member `G` different than `T.(I.G)`. Assuming it succceeds, though, it will
+    definitely have meaning determined by `I.G`. There may still be ambiguity
+    making the result dependent if it is not known whether `x` is a type and
+    `I.G` is a method and so has an implicit `me` parameter.
+-   If the member name is not found in the constraint, lookup may still succeed
+    once the type is known, so the result is template dependent.
+-   If the lookup is ambiguous prior to knowing the value of the type, for
+    example if `G` has two distinct meanings in `I`, then the code is invalid.
+
+The value of the expression will be dependent. For example, if `U` is an
+associated type of `I`, then `T.U` as an expression is template validity
+dependent, but the value of that expression is dependent. The value is not
+always needed to perform checking, for example `x.G()` can be checked without
+ever determining the value of `x.G` as long as its signature can be determined,
+if it is valid.
+
+#### Qualified member name lookup
+
+Adding qualifier to the member name, as in `T.(U.X)`, can make the lookup less
+dependent on the template parameter:
+
+-   If `T` is dependent, and `U` is an interface, the value of the expression is
+    dependent, but the type of the expression is not.
+    -   If `T` is known to implement `U`, then the lookup is not dependent, see
```

Comment:
I think this should now be... "If the type of `T` is known to implement `U`, ..." after #2360.